In [1]:
#This code block imports all the necessary libraries you need
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
#Import the csv file
df = pd.read_csv("/content/energydata_complete.csv")

In [3]:
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the Root Mean Squared error in three D.P?

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [5]:
#Selecting the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6)
X = df[['T2']]
y = df['T6']

In [6]:
#dividing the data (80% train, 20% test) into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [7]:
# Setting up the model for linear regression
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [8]:
# Predicting the test set
y_pred = model.predict(X_test)

In [9]:
#To find RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

In [10]:
#Print RMSE in 3 decimal places
print("Root Mean Squared Error:", "{:.3f}".format(rmse))

Root Mean Squared Error: 3.633


TASK: Remove the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler fit_transform and transform methods on the train and test set respectively). Run a multiple linear regression using the training set. Answer the following questions:

1. What is the Mean Absolute Error (in three decimal places) for the  training set?


In [11]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_absolute_error

In [12]:
#Drop column date and lights
df = df.drop(columns=['date', 'lights'])

In [13]:
# Specify the target variable (y) and characteristics (X).
X = df.drop(columns=['Appliances'])
y = df['Appliances']

In [14]:
##dividing the data (70% train, 30% test) into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [15]:
# Normalize using MinMaxScaler
scaler = MinMaxScaler()
X_train_normalized = scaler.fit_transform(X_train)
X_test_normalized = scaler.transform(X_test)

In [16]:
# Assign the training set to a multiple linear regression model
model = LinearRegression()
model.fit(X_train_normalized, y_train)

LinearRegression()

In [17]:
# Predicting the training set
y_train_pred = model.predict(X_train_normalized)

In [18]:
# Find Mean Absolute Error for the training set
mae_train = mean_absolute_error(y_train, y_train_pred)

In [19]:
#Print MAE
print("Mean Absolute Error for the training set:", "{:.3f}".format(mae_train))

Mean Absolute Error for the training set: 53.742


2. What is the Root Mean Squared Error (in three decimal places) for the training set?

In [20]:
from sklearn.metrics import mean_squared_error

In [21]:
# To first get the Mean Squared Error for the training set
mse_train = mean_squared_error(y_train, y_train_pred)

In [22]:
# Calculate Root Mean Squared Error for the training set
rmse_train = np.sqrt(mse_train)

In [23]:
#Printing RMSE
print("Root Mean Squared Error for the training set:", "{:.3f}".format(rmse_train))

Root Mean Squared Error for the training set: 95.216


In [24]:
# Predicting the test set
y_test_pred = model.predict(X_test_normalized)

In [26]:
# Finding MAE
mae_test = mean_absolute_error(y_test, y_test_pred)

In [27]:
#Print MAE
print("Mean Absolute Error for the test set:", "{:.3f}".format(mae_test))

Mean Absolute Error for the test set: 53.643


3. What is the Root Mean Squared Error (in three decimal places) for test set?

In [28]:
#MSE for the test set
mse_test = mean_squared_error(y_test, y_test_pred)

In [29]:
#RMSE for the test set
rmse_test = np.sqrt(mse_test)

In [30]:
#Print RMSE
print("Root Mean Squared Error for the test set:", "{:.3f}".format(rmse_test))

Root Mean Squared Error for the test set: 93.640


Train a ridge regression model with default parameters. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [31]:
from sklearn.linear_model import Ridge

In [32]:
# Training with default parameters
ridge_model = Ridge()
ridge_model.fit(X_train_normalized, y_train)

Ridge()

In [33]:
# Predicting the test set
y_test_pred_ridge = ridge_model.predict(X_test_normalized)

In [34]:
#The RMSE
rmse_test_ridge = np.sqrt(mean_squared_error(y_test, y_test_pred_ridge))

In [35]:
#Print RMSE
print("Root Mean Squared Error for test set with Ridge regression:", "{:.3f}".format(rmse_test_ridge))

Root Mean Squared Error for test set with Ridge regression: 93.709


Since RMSE values obtained from the Ridge regression model (93.709) and the RMSE values obtained from the previous linear regression model (93.640) are similar, it suggests that the Ridge regression model did not significantly improve or degrade the model's performance compared to the linear regression model.

4. Train a lasso regression model with default value and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [36]:
from sklearn.linear_model import Lasso

In [37]:
# Training with default parameters
lasso_model = Lasso()
lasso_model.fit(X_train_normalized, y_train)

Lasso()

In [46]:
#To get the feature weights
feature_weights = lasso_model.coef_
feature_weights

array([  0.        ,  56.94451601,   0.        ,  -0.        ,
         0.        ,   0.        ,  -0.        ,   0.        ,
        -0.        ,   0.        ,   0.        ,  -0.        ,
        -0.        ,  -0.        ,   0.        , -26.10288797,
        -0.        ,  -0.        ,   0.        ,  -0.        ,
       -50.29397595,   5.14314792,   0.        ,  -0.        ,
        -0.        ,  -0.        ])

In [39]:
# To count the number of features with non-zero feature weights
non_zero_feature_count = sum(feature_weights != 0)

In [40]:
#Print the number of features with non-zero feature weights
print("Number of features with non-zero feature weights:", non_zero_feature_count)

Number of features with non-zero feature weights: 4


5.  What is the new RMSE with the Lasso Regression on the test set?

In [41]:
#Predicting the test set with Lasso regression
y_test_pred_lasso = lasso_model.predict(X_test_normalized)

In [42]:
# The Root Mean Squared Error for the test set with Lasso regression
rmse_test_lasso = np.sqrt(mean_squared_error(y_test, y_test_pred_lasso))

In [43]:
#Print RMSE for the test set with Lasso regression
print("Root Mean Squared Error for test set with Lasso regression:", "{:.3f}".format(rmse_test_lasso))

Root Mean Squared Error for test set with Lasso regression: 99.424
